golang/cosmos/x/swingset/keeper/snapshotter.go

```diff
@@ -0,0 +1,216 @@
+package keeper
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/Agoric/agoric-sdk/golang/cosmos/vm"
+	"github.com/Agoric/agoric-sdk/golang/cosmos/x/swingset/types"
+	vstoragetypes "github.com/Agoric/agoric-sdk/golang/cosmos/x/vstorage/types"
+	"github.com/cosmos/cosmos-sdk/baseapp"
+	snapshots "github.com/cosmos/cosmos-sdk/snapshots/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/tendermint/tendermint/libs/log"
+	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+)
+
+var _ snapshots.ExtensionSnapshotter = &SwingsetSnapshotter{}
+
+const SnapshotFormat = 1
+
+type activeSnapshot struct {
+	height        int64
+	startedResult chan error
+	retrieved     bool
+}
+
+type SwingStoreExporter interface {
+	ExportSwingStore(ctx sdk.Context) []*vstoragetypes.DataEntry
+}
+
+type SwingsetSnapshotter struct {
+	app            *baseapp.BaseApp
+	exporter       SwingStoreExporter
+	blockingSend   func(action vm.Jsonable) (string, error)
+	activeSnapshot *activeSnapshot
+}
+
+type snapshotAction struct {
+	Type        string            `json:"type"` // COSMOS_SNAPSHOT
+	BlockHeight int64             `json:"blockHeight"`
+	Request     string            `json:"request"` // "initiate", "discard", "retrieve", or "restore"
+	Args        []json.RawMessage `json:"args,omitempty"`
+}
+
+func NewSwingsetSnapshotter(app *baseapp.BaseApp, exporter SwingStoreExporter, sendToController func(bool, string) (string, error)) SwingsetSnapshotter {
+	blockingSend := func(action vm.Jsonable) (string, error) {
+		bz, err := json.Marshal(action)
+		if err != nil {
+			return "", err
+		}
+		return sendToController(true, string(bz))
+	}
+
+	return SwingsetSnapshotter{
+		app:            app,
+		exporter:       exporter,
+		blockingSend:   blockingSend,
+		activeSnapshot: nil,
+	}
+}
+
+func (snapshotter *SwingsetSnapshotter) logger() log.Logger {
+	return snapshotter.app.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName), "submodule", "snapshotter")
+}
+
+func (snapshotter *SwingsetSnapshotter) InitiateSnapshot(height int64) error {
+	if snapshotter.activeSnapshot != nil {
+		return fmt.Errorf("snapshot already in progress for height %d", snapshotter.activeSnapshot.height)
+	}
+
+	if snapshotter.app.SnapshotManager() == nil {
+		return fmt.Errorf("snapshot manager not configured")
+	}
+
+	snapshotter.activeSnapshot = &activeSnapshot{
+		height,
+		make(chan error, 1),
+		false,
+	}
+
+	go func(startedResult chan error) {
+		logger := snapshotter.logger()
+
+		action := &snapshotAction{
+			Type:        "COSMOS_SNAPSHOT",
+			BlockHeight: snapshotter.activeSnapshot.height,
+			Request:     "initiate",
+		}
+
+		_, err := snapshotter.blockingSend(action)
```

Comment:
Still need to see this particular aspect documented. Here's a suggestion - please correct the details if they're not right:

```
// IMPORTANT NOTE: this bridge message is sent by the goroutine, so it might be sent in different orders
// on different replicas. This is safe because:
// 1) The effects of this message do not affect the application hash and thus are outside of consensus.
// In particular, it does not affect GC, computation metrics, etc., in a way that would affect the processing
// of consensus-relevant events.
// 2) The artifacts produced depend only on data settled in previous block heights and is unaffected by
// bridge messages which might occur in different orders.
```

Reply:
I addressed the non-determinism in the definition of `snapshotter.blockingSend` here:
https://github.com/Agoric/agoric-sdk/pull/7225/commits/8b471f117f6d0c2724aaab8f44cc8afc610fb339#diff-1b51e864e9c6f66ee4ca82b7a035189ff2cf5e288621682aeef380104eea4a8dR65-R69

That comment does not incorporate the point under 2), however I'm not sure it belongs under documentation for `sendToController`, as it relates more to the general synchronization that the snapshotter performs.